### PR TITLE
NativeAOT fixes around data flow analysis

### PR DESCRIPTION
**Pull request by vitek-karas**
Removes the `Origin` class as it doesn't seem to serve any purpose anymore. The code used it to basically just only call `GetDisplayName` on members and pass that as the `reason` to dependency analysis. Replaced with passing just string directly as the `reason`.

Fixes bugs in warning subsystem when the warning is generated from an origin which is a module or assembly. This would previously crash the compiler. There are basically no tests for this right now - some of the tests in linker which hit this can't be easily ported yet due to NativeAOT not producing "reflection access to DAM" warnings. I have these mostly working locally but enabling the warnings fails some tests outside of the linker ones, so it needs more work.

Fixes analysis bugs when methods or fields are accessed via `ldtoken` or `ldftn` instructions. In such cases we still need to process them for RUC and DAM based warnings and so on. There are only a few tests for this in this change - more will come when we port the `AnnotatedMembersAccessedViaReflection` from linker, but those all require to enable the "reflection access to DAM" warnings as mentioned above.
* This required addition of a new TrimAnalysisPattern since the processing happens in reflection scanning and all warnings/marking hapening from there must go through the patterns.

Changes `DynamicDependencyAttribute` processing to happen in a separate dependency node. Previously if there were warnings generated due to the `DynamicDependencyAttribute` the compiler may report them multiple times (if it hit the same member twice during the rest of the analysis). To avoid this, the processing is postponed by adding a new dependency node - which then acts as the unification mechanism and thus warnings are produced just once.

### CODE REVIEW

File: src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs [hunk 1]
```diff
@@ -954,6 +957,9 @@ private void ImportLdToken(int token)
             else
             {
                 Debug.Assert(obj is FieldDesc);
+                var field = (FieldDesc)obj;
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:960]: The assert is now of little value.

```suggestion
                var field = (FieldDesc)obj;
```

**Review by MichalStrehovsky — APPROVED** [review]
Thank you!

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs [hunk 2]
```diff
@@ -693,7 +693,7 @@ public override void GetDependenciesDueToAccess(ref DependencyList dependencies,
             string reason = "Use of a field";
 
             bool generatesMetadata = false;
-            if (!IsReflectionBlocked(writtenField))
+            if (!writtenField.OwningType.IsRuntimeDeterminedSubtype && !IsReflectionBlocked(writtenField))
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs:696]: It's usually a bug if runtime determined forms reach here. Someone should have converted to canonical. I think we might miss things if we treat these as blocked.

Comment (vitek-karas) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs]: Thanks for pointing this out - fixed by passing only canonical forms from the ILScanner.

**Pull request merged**